Cannot use `lisette.MakeTuple2(...)` as `lisette.Tuple2[...]`

Given the following lisette program:
```rs
import "go:charm.land/bubbletea/v2"
import uv "go:github.com/charmbracelet/ultraviolet" // this is only here to temporarily fix the error reported in issue #99

struct Model { str: string }

impl Model {
  fn Init(self: Model) -> Option<tea.Cmd> {
    return None
  }

  fn Update(self: Model, _msg: tea.Msg) -> (tea.Model, Option<tea.Cmd>) {
    return (self, None)
  }

  fn View(self: Model) -> tea.View {
    return tea.NewView(self.str)
  }
}

fn main() {
  let _ = tea.NewProgram(Model { str: "hello, world" }).Run()
}
```
I get the following error from the generated `Go` code when doing `lis run`:
> ./main.go:23:9: cannot use lisette.MakeTuple2(m, lisette.MakeOptionNone\[tea.Cmd\]()) (value of struct type lisette.Tuple2[Model, lisette.Option[tea.Cmd]]) as lisette.Tuple2[tea.Model, lisette.Option[tea.Cmd]] value in return statement

That error comes from this part of the generated Go file:
```go
func (m Model) Update(_ uv.Event) lisette.Tuple2[tea.Model, lisette.Option[tea.Cmd]] {
	return lisette.MakeTuple2(m, lisette.MakeOptionNone[tea.Cmd]()) // this line is where the above error diagnostic shows up
}
```

--- a/crates/emit/src/go/control_flow/propagation.rs
+++ b/crates/emit/src/go/control_flow/propagation.rs
@@ -189,7 +189,8 @@ impl Emitter<'_> {
             }
             output.push_str("return\n");
         } else if !self.emit_wrapped_return(output, expression) {
-            let expression_string = self.emit_value(output, expression);
+            let expression_string =
+                self.with_position(Position::Tail, |this| this.emit_value(output, expression));
             let expression_string = self.adapt_return_to_context(expression, expression_string);
             write_line!(output, "return {}", expression_string);
         }

--- a/crates/emit/src/go/expressions/access.rs
+++ b/crates/emit/src/go/expressions/access.rs
@@ -657,7 +657,7 @@ impl Emitter<'_> {
     }
 
     fn add_enum_imports_if_needed(&mut self, name: &str, enum_id: &str) {
-        let enum_module = enum_id.split('.').next().unwrap_or("");
+        let enum_module = go_name::module_of_type_id(enum_id);
 
         if enum_module != self.current_module {
             self.require_module_import(enum_module);

--- a/crates/emit/src/go/expressions/dot_classify.rs
+++ b/crates/emit/src/go/expressions/dot_classify.rs
@@ -24,7 +24,7 @@ impl Emitter<'_> {
             _ => return None,
         };
 
-        let (module_key, _type_name) = enum_id.split_once('.')?;
+        let module_key = go_name::module_of_type_id(&enum_id);
 
         let qualifier = self.go_pkg_qualifier(module_key);
 
@@ -104,7 +104,7 @@ impl Emitter<'_> {
 
         let make_fn_name = self.module.make_functions.get(&constructor_key)?.clone();
 
-        let enum_module = enum_id.split('.').next().unwrap_or("");
+        let enum_module = go_name::module_of_type_id(enum_id);
         let needs_qualifier = enum_module != self.current_module;
 
         let needs_type_args = ret_params.len() > fn_params.len();

--- a/crates/emit/src/go/expressions/identifiers.rs
+++ b/crates/emit/src/go/expressions/identifiers.rs
@@ -346,9 +346,10 @@ impl Emitter<'_> {
         }
 
         let variant_name = go_name::unqualified_name(name);
-        let module = enum_id.split('.').next()?;
+        let module = go_name::module_of_type_id(enum_id.as_str());
+        let qualifier = self.go_pkg_qualifier(module);
 
-        Some(format!("{}.{}", module, variant_name))
+        Some(format!("{}.{}", qualifier, variant_name))
     }
 
     /// Check if an identifier refers to a public function in the current module.

--- a/crates/emit/src/go/expressions/values.rs
+++ b/crates/emit/src/go/expressions/values.rs
@@ -245,10 +245,24 @@ impl Emitter<'_> {
                     elements.iter().map(|e| self.stage_composite(e)).collect();
                 let elem_expressions = self.sequence(output, stages, "_v");
 
-                let slot_types: Vec<Type> = match ty.resolve() {
+                let inferred_slot_types: Vec<Type> = match ty.resolve() {
                     Type::Tuple(slots) => slots,
                     _ => Vec::new(),
                 };
+
+                let slot_types: Vec<Type> = self
+                    .position
+                    .is_tail()
+                    .then_some(self.current_return_context.as_ref())
+                    .flatten()
+                    .and_then(|ret| {
+                        if let Type::Tuple(slots) = ret.resolve() {
+                            (slots.len() == inferred_slot_types.len()).then_some(slots)
+                        } else {
+                            None
+                        }
+                    })
+                    .unwrap_or(inferred_slot_types);
 
                 let elem_expressions: Vec<String> = elements
                     .iter()

--- a/crates/emit/src/go/names/go_name.rs
+++ b/crates/emit/src/go/names/go_name.rs
@@ -43,6 +43,16 @@ pub(crate) fn go_package_name(module: &str) -> &str {
     module.rsplit('/').next().unwrap_or(module)
 }
 
+pub(crate) fn module_of_type_id(id: &str) -> &str {
+    if let Some(slash_pos) = id.rfind('/') {
+        let after_slash = slash_pos + 1;
+        if let Some(dot_offset) = id[after_slash..].find('.') {
+            return &id[..after_slash + dot_offset];
+        }
+    }
+    id.split('.').next().unwrap_or(id)
+}
+
 pub(crate) fn sanitize_package_name(name: &str) -> Cow<'_, str> {
     let has_bad_chars = name.chars().any(|c| !c.is_ascii_alphanumeric() && c != '_');
     let starts_with_digit = name.starts_with(|c: char| c.is_ascii_digit());
@@ -161,7 +171,7 @@ pub(crate) fn variant_by_id(
     module_alias: Option<&str>,
 ) -> ResolvedName {
     let is_prelude = enum_id.starts_with(PRELUDE_PREFIX);
-    let enum_module = enum_id.split('.').next().unwrap_or("");
+    let enum_module = module_of_type_id(enum_id);
     let enum_name = enum_id.split('.').next_back().unwrap_or(enum_id);
     let variant_name = identifier.split('.').next_back().unwrap_or(identifier);
 

--- a/crates/emit/src/go/patterns/decision_tree.rs
+++ b/crates/emit/src/go/patterns/decision_tree.rs
@@ -693,6 +693,25 @@ fn collect_enum_variant_checks(
         return;
     }
 
+    if emitter.is_go_value_enum(ty) {
+        let Type::Constructor { id, .. } = ty.resolve().strip_refs() else {
+            return;
+        };
+        let variant_name = go_name::unqualified_name(identifier);
+        let module = go_name::module_of_type_id(id.as_str());
+        let qualifier = emitter.go_pkg_qualifier(module);
+        let go_literal = if qualifier.is_empty() || qualifier == emitter.current_module() {
+            variant_name.to_string()
+        } else {
+            format!("{}.{}", qualifier, variant_name)
+        };
+        collector.checks.push(Check::Literal {
+            path: path.clone(),
+            go_literal,
+        });
+        return;
+    }
+
     collect_tagged_enum_checks(emitter, path, &variant_data, collector);
 }
 

--- a/crates/emit/src/go/queries/definitions.rs
+++ b/crates/emit/src/go/queries/definitions.rs
@@ -124,6 +124,16 @@ impl Emitter<'_> {
         )
     }
 
+    pub(crate) fn is_go_value_enum(&self, ty: &Type) -> bool {
+        let Type::Constructor { id, .. } = ty.resolve().strip_refs() else {
+            return false;
+        };
+        matches!(
+            self.ctx.definitions.get(id.as_str()),
+            Some(Definition::ValueEnum { .. })
+        )
+    }
+
     pub(crate) fn get_newtype_underlying(&self, ty: &Type) -> Option<Type> {
         let Type::Constructor { id, .. } = ty.resolve().strip_refs() else {
             return None;

--- a/crates/emit/src/lib.rs
+++ b/crates/emit/src/lib.rs
@@ -407,7 +407,7 @@ impl<'a> Emitter<'a> {
 
     pub(crate) fn module_alias_for_type(&self, ty: &Type) -> Option<String> {
         if let Type::Constructor { id, .. } = ty {
-            let module = id.split('.').next().unwrap_or("");
+            let module = go::names::go_name::module_of_type_id(id);
             self.module.module_aliases.get(module).cloned()
         } else {
             None

--- a/tests/spec/emit/interop_matrix.rs
+++ b/tests/spec/emit/interop_matrix.rs
@@ -827,6 +827,85 @@ fn main() {
 }
 
 #[test]
+fn interop_value_enum_match_arm_aliased() {
+    let input = r#"
+import t "go:time"
+
+fn describe(d: t.Duration) -> string {
+  match d {
+    t.Duration.Second => "one second",
+    t.Duration.Minute => "one minute",
+    _ => "other",
+  }
+}
+"#;
+    assert_emit_snapshot!(input);
+}
+
+#[test]
+fn interop_value_enum_match_arm_nested_module() {
+    let input = r#"
+import "go:debug/dwarf"
+
+fn describe(a: dwarf.Attr) -> string {
+  match a {
+    dwarf.Attr.AttrArtificial => "artificial",
+    dwarf.Attr.AttrByteSize   => "byte size",
+    _ => "other",
+  }
+}
+"#;
+    assert_emit_snapshot!(input);
+}
+
+#[test]
+fn interop_tuple_tail_concrete_in_go_interface_slot() {
+    let input = r#"
+import "go:fmt"
+
+struct Counter {
+  count: int,
+}
+
+impl Counter {
+  fn String(self) -> string {
+    f"{self.count}"
+  }
+}
+
+fn make_pair(c: Counter) -> (fmt.Stringer, int) {
+  (c, c.count)
+}
+"#;
+    assert_emit_snapshot!(input);
+}
+
+#[test]
+fn interop_tuple_explicit_return_concrete_in_go_interface_slot() {
+    let input = r#"
+import "go:fmt"
+
+struct Counter {
+  count: int,
+}
+
+impl Counter {
+  fn String(self) -> string {
+    f"{self.count}"
+  }
+}
+
+fn make_pair(c: Counter, positive: bool) -> (fmt.Stringer, int) {
+  if positive {
+    return (Counter { count: c.count + 1 }, c.count)
+  }
+  (c, c.count)
+}
+"#;
+    assert_emit_snapshot!(input);
+}
+
+#[test]
 fn interop_aliased_import_type_reference() {
     let input = r#"
 import t "go:time"

